### DISCUSSION TITLE
[창훈] assignment 기능구현

### DISCUSSION THREAD

**Discussion by anotheranotherhoon** [Wanted-07-team-9/pre-onboarding-7th-2-1-9#18] [기능구현]
![React](https://img.shields.io/badge/react-%2320232a.svg?style=for-the-badge&logo=react&logoColor=%2361DAFB) ![NPM](https://img.shields.io/badge/NPM-%23000000.svg?style=for-the-badge&logo=npm&logoColor=white) ![badge](https://img.shields.io/badge/tailwindcss-%2338B2AC.svg?style=for-the-badge&logo=tailwind-css&logoColor=white) ![Vercel](https://img.shields.io/badge/vercel-%23000000.svg?style=for-the-badge&logo=vercel&logoColor=white) ![Recoil](https://img.shields.io/badge/recoil-f26b00?style=for-the-badge&logo={https://relay.dev/img/relay.svg})

> pre-onboarding-7th-2-1-9

## **Description**

차량 대여를 위해 차량 목록을 불러와서 표시해주는 서비스입니다.  
상태관리 라이브러리로 Recoil을 css라이브러리로 tailwind를 사용하였습니다. 

## **Demo Page**

[배포 페이지](https://pre-onboarding-7th-2-1-9-eta.vercel.app/)

## **Assignment 1**

- Figma 상의 디자인 및 기능 구현
  - 모바일 웹 기준으로 제작
  - **450px ~ 360px**까지 고려해서 제작
![스크린샷 2022-11-03 오후 9 47 10](https://user-images.githubusercontent.com/77476348/199724200-bf7653a0-3e9b-4e30-8136-20317772db0d.png)

- 필수 요구 사항
  - Figma 상의 디자인 및 기능 구현

## **Assignment 2**

https://github.com/Wanted-07-team-9/pre-onboarding-7th-2-1-9/blob/79faf69de3aebc1fc26695c4b8861adb8f14a66d/src/container/Main.js#L10-L33

- 차량 리스트
```jsx
  const segment = useRecoilValue(segmentParameter)
  const CarsLoadable = useRecoilValueLoadable(carList(segment))
```
segment의 값이 바뀔 때마다 차량리스트를 GET하는 함수에 파라미터를 바꿔준다. useEffect에 의존성배열에 segment를 넣어준 것 처럼 작동하며
segment의 초기 값이 null 이므로 최초에는 모든 차량  리스트를 가져온다. 

- 차량이 없을 때 처리
```jsx
if (Cars.length === 0) {
content = <PageState state={'empty'} />
}
```
차량리스트에서 빈 배열일 경우 차량 데이터가 없다는 메세지를 보여준다. 

- 차량 불러오는 중 처리

```jsx
// CarsLoadable은 hasValue, hasError, loading 상태를 가진다. 
case 'loading':
  content = <Loading />
```
loading 상태일 경우 Loading 컴포넌트를 보여준다. 

## **Assignment 3**

- 차랑 상세

![스크린샷 2022-11-03 오후 9 47 10](https://user-images.githubusercontent.com/77476348/199724200-bf7653a0-3e9b-4e30-8136-20317772db0d.png)

https://github.com/Wanted-07-team-9/pre-onboarding-7th-2-1-9/blob/79faf69de3aebc1fc26695c4b8861adb8f14a66d/src/container/CarDetail.js#L14

https://github.com/Wanted-07-team-9/pre-onboarding-7th-2-1-9/blob/79faf69de3aebc1fc26695c4b8861adb8f14a66d/src/recoil/carList.js#L12-L19

```jsx
  const { state } = useLocation()

const handleClickCarInfo = () => {
    navigate(`/car/${car.id}`, {
      state: car
    })
  } 
```

차량의 id로 차량 하나의 데이터를 가져오는 API가 없어서 차량 클릭시 /car/${car.id} 로 네비게이트 하면서 두 번째 인자로 차량의 정보를 함께 보내주려고 했으나 클릭이 아닌 해당 URL에 직접 이동하려 할 경우 SEO 이슈가 발생할 것이라 판단되어, 전체 데이터에서 데이터를 필터하는 방법을 채택하였다.
차량 전체 리스트를 가져오는 useRecoilValueLoadable의 로직과 같다. 

## **Assignment 4 (추가 구현 사항)**

- **SEO 구현**
https://github.com/Wanted-07-team-9/pre-onboarding-7th-2-1-9/blob/79faf69de3aebc1fc26695c4b8861adb8f14a66d/src/utils/setMetaTags.js#L1-L15

![스크린샷 2022-11-03 오후 10 19 15](https://user-images.githubusercontent.com/77476348/199731251-080ce1bc-b642-4edd-90b0-0599645f84e6.png)
![](https://velog.velcdn.com/images/anotherhoon/post/e2a0603e-e954-432c-b91f-23edff0a32e2/image.png)
메타 태그를 동적으로 바꾸는 함수를 이용하였고 개발자 도구에서는 잘 적용되었으나, SNS에서는 적용하지 못 했다.

**Comment by tirhande** [comment]
Assignment 2
- recoil 쓰셔서 상태 관리 하신점이 좋네요!!! 👍
- 그리고 switch로 컴포넌트 변경하는것도 참신해서 좋아요!

**Comment by jun-05** [comment]
- tailwind나 recoil 등, 새로운거 찾아서 쓰시는게 대단하신거 같아요.
- 전체적으로 잘 구현하신거 같습니다!
- 아쉬운건 CSS를 max-width라도 잡아줬으면 어땠을까 하는 생각이 있었네요.

**Comment by songseul** [comment]
- recoil로 상태 관리 하는 점이 인상 깊어요!! 
- switch를 사용 하시거나 필터로 값을 뿌리는 방법들 이전 과제에서 사용 하신 방법 말도로 시도 하시는 점이 좋은 방향 같습니다
- 피씨에서도 봤을때 전체 너비를 잡아 주시면 좋을것 같아요

**Comment by im-binary** [comment]
- 차량을 불러올 때나 없을 때 에러나 switch 문으로 처리를 잘 하신 것 같습니다!
-  recoil을 쓰니까 확실히 타 상태관리 라이브러리보다 가독성이 좋은 것 같네용
- SEO 구현하기 위한 노력 멋있으십니다..!

아 궁금한 게 있는데,
```js
 let Cars = [] 
```
배열을 선언할 때 let으로 선언하고 변수 이름이 대문자로 시작하는 이유가 있을까요?!

**Reply by anotheranotherhoon** [reply]
car 로 바꾸는 편이 좋을 것 같네요! 감사합니다.

**Comment by NR0617** [comment]
차량이 없을때 state를 props로 넘겨주는 방법을 사용한게 재미있어요
switch case문을 사용하기 적절한 부분이라고 생각합니다 저도 다음에 사용해보고싶어요!
